File: modules/cubex-core/src/main/java/org/cubexmc/core/TerminableRegistry.java
package org.cubexmc.core;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Objects;

final class TerminableRegistry {

    private final Deque<AutoCloseable> terminables = new ArrayDeque<>();

    <T extends AutoCloseable> T bind(T terminable) {
        terminables.addLast(Objects.requireNonNull(terminable, "terminable"));
        return terminable;
    }

    Terminable bind(Runnable closeAction) {
        return bind(Terminable.of(closeAction));
    }

    void closeAll(CloseFailureHandler failureHandler) {
        while (!terminables.isEmpty()) {
            AutoCloseable terminable = terminables.removeLast();
            try {
                terminable.close();
            } catch (Exception ex) {
                failureHandler.handle(ex);
            }
        }
    }

    @FunctionalInterface
    interface CloseFailureHandler {
        void handle(Exception exception);
    }
}

File: modules/cubex-core/src/main/kotlin/org/cubexmc/core/TerminableRegistry.kt
package org.cubexmc.core

import java.util.ArrayDeque
import java.util.Deque
import java.util.Objects

/** A LIFO resource group for a plugin or one of its domain managers. */
class TerminableRegistry {
    private val terminables: Deque<AutoCloseable> = ArrayDeque()

    fun <T : AutoCloseable> bind(terminable: T): T {
        terminables.addLast(Objects.requireNonNull(terminable, "terminable"))
        return terminable
    }

    fun bind(closeAction: Runnable): Terminable = bind(Terminable.of(closeAction))

    fun closeAll(failureHandler: CloseFailureHandler) {
        while (terminables.isNotEmpty()) {
            val terminable = terminables.removeLast()
            try {
                terminable.close()
            } catch (exception: Exception) {
                failureHandler.handle(exception)
            }
        }
    }

    fun interface CloseFailureHandler {
        fun handle(exception: Exception)
    }
}
